cmd/controller/collectdiskinfo.go

```diff
@@ -0,0 +1,98 @@
+/*
+Copyright 2018 OpenEBS Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"io/ioutil"
+	"strconv"
+
+	apis "github.com/openebs/node-disk-manager/pkg/apis/ndm/v1alpha1"
+	"github.com/openebs/node-disk-manager/pkg/udev"
+)
+
+/*
+collectcollectDiskInfo will collect info about a disk from different
+probe like udev a sample return api.Disk object is given bellow -
+v1alpha1.Disk{
+	TypeMeta:v1.TypeMeta{
+		Kind:"Disk",
+		APIVersion:"openebs.io/v1alpha1"
+		},
+		ObjectMeta:v1.ObjectMeta{
+			Name:"disk-0f2f3ce231cfe360489655c5d928b514"
+		},
+		Spec:v1alpha1.DiskSpec{
+			Path:"/dev/sda",
+			Capacity:v1alpha1.DiskCapacity{
+				Storage:0xe8e0db6000
+			},
+			Details:v1alpha1.DiskDetails{
+				Model:"ST1000LM024_HN-M101MBB",
+				Serial:"S31LJ9JG803819",
+				Vendor:"ATA"
+			}
+		},
+		Status:v1alpha1.DiskStatus{
+			State:"Active"
+		}
+	}
+*/
+func collectDiskInfo(device *udev.Udevice, hostName string) (apis.Disk, error) {
+	dr := apis.Disk{}
+	var sector []byte
+	var sec int64
+	uuid := GetUid(device)
+	udevDiskDetails := diskInfoFromLibudev(device)
+	n, err := strconv.ParseInt(device.SysattrValue("size"), 10, 64)
+	if err != nil {
+		return dr, err
+	}
+	// should we use disk smart queries to get the sector size?
+	fname := "/sys" + udevDiskDetails[udev.UDEV_PATH] + "/queue/hw_sector_size"
+	sector, err = ioutil.ReadFile(fname)
+	if err != nil {
+		return dr, err
+	}
+	sec, err = strconv.ParseInt(string(sector[:len(sector)-1]), 10, 64)
+	if err != nil {
+		return dr, err
+	}
+	obj := apis.DiskSpec{Path: device.Devnode()}
+	obj.Capacity.Storage = uint64(n * sec)
+	obj.Details.Model = udevDiskDetails[udev.UDEV_MODEL]
+	obj.Details.Serial = udevDiskDetails[udev.UDEV_SERIAL]
+	obj.Details.Vendor = udevDiskDetails[udev.UDEV_VENDOR]
+	dr.ObjectMeta.Labels = make(map[string]string)
+	dr.ObjectMeta.Labels[NDMHostKey] = hostName
+	dr.TypeMeta.Kind = NDMKind
+	dr.TypeMeta.APIVersion = NDMVersion
+
+	dr.Spec = obj
+	dr.Status.State = NDMActive
+	dr.ObjectMeta.Name = uuid
+	return dr, nil
+}
+
+//diskInfoFromLibudev returns disk attribute extracted using libudev apicalls.
+func diskInfoFromLibudev(device *udev.Udevice) map[string]string {
```

Comment:
Just wondering if this can be a method of *udev.Udevice?
This function i.e. diskInfoFromLibudev & other function i.e. collectDiskInfo can be merged & be made available as a method of *udev.Udevice.

e.g.
```go
type DeviceInformer interface {
  DiskInfo() DiskInfo
}

func (d *udev.Udevice) DiskInfo() DiskInfo {
...
}
```

How are we going to get disk info from devices which are not udev based. I may not be putting the real device terms, but IMO it will be good to have dedicated interfaces & structures to manage the functionality.

Reply:
made some changes using struct.